### ISSUE TITLE
MiMo observer wiring + one-pass REAP/HOPE observation run (1% then convergence-sized)

### ISSUE THREAD

**Issue by dymoo** [dymoo/mimo-halo-lab#78]
Part of #76 · Mission: #1

## Goal

Wire the MiMo model attributes + observer registry into the REAP observer (`src/mimo_halo/pruning/reap_adapter.py`) for one-pass REAP+HOPE observation, then execute the one-pass observation run: a 1% calibration-corpus smoke first, then the convergence-sized streamed run. This is the ONLY sanctioned way to produce fresh observation statistics — no reuse of tacodevs numbers.

## Inputs

- Pinned MiMo tensor inventory: `manifests/models/mimo-v2.6-flash-rl/{inventory.json,memory-report.json}` (verified; `tests/test_inventory.py` 33/33).
- Upstream REAP pin `1970473c` per `manifests/upstreams.json` (all 4 upstreams HEAD==pin).
- Observation engine: resumable layer-wise observation with atomic checkpoints — issue #21 (open).
- Calibration corpus from our local trace partitions (pruning 922 / quant 486 task groups; golden 235 excluded by rule).
- HOPE objective + validation already landed: `src/mimo_halo/pruning/hope.py`, `docs/hope-objective.md` (closed #17, closed #18).
- In-flight agent **MiMoObserverWire** owns `src/mimo_halo/pruning/reap_adapter.py` (uncommitted at capture) — KNOWN OPEN GAP in the wiring currently blocks real REAP selection; land the wiring first.

## Outputs

- One-pass observation artifacts: per-layer REAP scores AND paper-faithful HOPE statistics (first/pairwise terms, frequency, mean gate, output norms) from the SAME pass, with resumable checkpoints.
- 1% run record: wall time, peak memory, parse/stat failure count (target 0).
- Convergence-sized run record with token counts at 50K/100K/250K/500K/1M boundaries (stability evidence consumed by #33).
- Provenance sidecars for every artifact (`src/mimo_halo/artifacts.py`).

## Blockers

- Blocked by: #21

## Acceptance

- [ ] MiMo MODEL_ATTRS/observer registry wired for the verified architecture (47 MoE layers 1–47 + dense layer 0, 256 experts, top-8) with an automated test or recorded verification run.
- [ ] 1% corpus observation completes with 0 stat failures; wall time + peak memory recorded.
- [ ] Convergence-sized streamed run completes (or resumes across interruption) without losing completed layers.
- [ ] Both REAP and HOPE statistics exist for the same pass — no second pass, no imported community statistics.
- [ ] Every artifact has a provenance sidecar (source commits, dataset manifest, checksums, seed, environment).

## Commands (contract targets — verify entry points before running)

```bash
PYTHONPATH=src python -m pytest tests/test_inventory.py tests/test_hope.py tests/test_prune_maps.py
# wiring smoke (entry point provided by the landed adapter):
PYTHONPATH=src python -m mimo_halo.pruning.reap_adapter --help
# 1% observation, then convergence-sized run (flags per docs/):
PYTHONPATH=src python -m mimo_halo.pruning.reap_adapter observe --corpus <calibration> --fraction 0.01
```

## Provenance

Observation manifest + checksums under `manifests/`; parent spec #76, mission #1.

**Comment by dymoo** [comment]
**Split status (2026-09-22): wiring DONE / observation run OPEN.**

- DONE — wiring half, landed as commit `4db03b5` "Wire MiMo into one-pass REAP+HOPE observation": adapter-owned MODEL_ATTRS/observer registry for `MiMoV2ForCausalLM`, pinned REAP clone untouched, fail-closed MXFP4 packing verification, augment/restore forward hooks recompute router logits; `tests/test_reap_observer.py` — 66 tests green.
- OPEN — observation-run half: the 1% corpus smoke and the convergence-sized streamed run have NOT been executed (issue stays open + `blocked` on #21). Outputs required by this ticket's acceptance remain outstanding.

**Comment by dymoo** [comment]
Arrival deadline priority: 10:00 BST Sep23 workable handoff, improvements by13:00. Current worker evidence: real-weight dense layer0 and MoE layer1 finite forward passes; REAP/HOPE first-order max delta 3.8e-6, routing frequency sum8, peak RSS9.55GB. Full47-layer smoke is still executing, not accepted yet. Its initial22,528-token corpus used several non-golden partitions and one capability: mechanics evidence ONLY, not representative or accepted pruning calibration. Before usable selection stats: rebuild diverse pruning-only corpus; exclude all reserved-golden members plus borrowed-verification evidence-source sessions from the private oracle registry; verify exclusion manifests. Obtain complete50K and100K all-layer checkpoints before attempting1M. A working fixture adapter, encoded weights or manifest integrity does not prove actual inference-loader compatibility; separate loadability probe is running. No quality winner or full observation completion claimed.

**Comment by dymoo** [comment]
Critical discovery for overnight readiness: initial real-weight smoke used incorrect fused-QKV FP8 dequantization (naive sequential scales instead of tensor-parallel rank-aware Q/K/V layout). Worker reports corrected path exactly matches pinned converter on real layers0/1/5 (max absolute diff0); old layer0 discrepancy reached1.668. Pre-fix smoke statistics are INVALID and must never drive pruning selections or quality claims; finite outputs alone were insufficient evidence. Directed worker to stop unfinished invalid smoke, preserve labeled mechanics logs, then rerun corrected mini/all-layer smoke and50K/100K checkpoints on exclusion-safe pruning-only corpus. Separate builder blocker discovered: output config retains original expert count after router slicing; bounded fix in progress. Inference-loadable candidate compatibility remains an explicit acceptance requirement, not established by fixture artifact tests.

**Comment by dymoo** [comment]
Corrected smoke2 checkpoint update: worker found no live process and no final run record; valid outputs exist through MoE layers1-16 (two segments each). Termination cause is UNPROVEN—no exit/counter record establishes OOM. Resumed from next_layer17 with microbatch4->2, halving CPU scratch17.2GB->8.6GB (calculated sizes), existing parts preserved. New instructions: persist actual per-layer MPS/CPU memory+elapsed telemetry, verify inference_mode/no_grad and layer-reference release, record changed execution config. Corpus excludes golden members/evidence sources and uses pruning only, but one long session supplied all34rows/34,816 tokens; this is mechanics-only evidence, NOT diverse calibration or quality-ranked REAP selection.50K corpus must enforce per-session caps and capability diversity before the next run.

**Comment by dymoo** [comment]
Full-layer mechanics finding: corrected smoke2 reached all47MoE layers and wrote94HOPE segments across two incarnations, but final REAP export failed with TypeError (Tensor count not JSON serializable). More importantly, REAP was only held in-process: resumed runs lost earlier-layer REAP, and final in-memory part lists omitted prior-incarnation files. This invalidates shared REAP/HOPE completeness despite successful HOPE mechanics. Stopped another single-session rerun; fixing the actual resume contract: JSON-safe per-layer REAP parts, partial-layer replay without duplicate HOPE, authoritative on-disk full-grid validation and agreement. Regression must interrupt after a completed layer and inside the next. After that: inspect diverse50K corpus, then run50K as all-layer acceptance. Original silent-stop cause remains unproven; recovered3600s launch timeout rules out its supervisor deadline. Hidden tensors are recomputed by prefix replay, not persisted.

**Comment by dymoo** [comment]
The first reworked 50K corpus is NOT accepted for REAP quality ranking: despite 56×1024=57,344 token rows and 3 capability labels (17 compiler/25 debugging/14 implementation), it contains only 3 source sessions and 3 episodes out of 963 eligible sessions. Declared per_session_token_cap=1024 was bypassed by atomic long episodes (17–25K rows contributed from single sources), so capability coverage filled the budget prematurely. No full observation launched on it; it remains a private invalid-representativeness mechanics artifact. Builder now enforcing a HARD one-complete-row-per-distinct-source cap and will report qualifying source/episode coverage before any run approval. Separate checkpoint work: partial-layer REAP+HOPE restart seam exercised, but still require prior-completed-layer + next-partial-layer real resume regression for final acceptance. Old incomplete runs cannot be resumed as quality evidence.

**Comment by dymoo** [comment]
Cross-layer process-restart integrity is now proven in the actual tiny streamed observer: process A completes MoE layer1 REAP+HOPE then checkpoints layer2 segment0; process B replays that segment to rebuild in-memory REAP and finishes layer2 segment1 without duplicating HOPE; disk scan recovers 2 REAP + 4 unique HOPE parts and both layers match uninterrupted control (HOPE ≤1e-12, REAP ≤1e-4). Main reran tests.test_observation_resume + tests.test_reap_observer: 28 OK. Prior incomplete smoke2 was rejected early for missing REAP parts and cannot be reused as quality data. The corpus v1 still fails representativeness (only3 source sessions); v2 hard one-complete-row-per-distinct-source build is in progress. No full-size observation has launched or measured quality.

**Comment by dymoo** [comment]
REAP calibration corpus accepted after rejecting concentrated v1: private corpus-50k-v2 is 49 complete 1024-token rows = 50,176 tokens (SHA256 rows.npy f407ec82a2b66f0465ff60b9e8a8be8289a6bcf3b9c87ba0ca521846c3766925), from 49 distinct source sessions, 49 episodes, and 49 distinct frozen pruning task groups; max1 row/source/group. 12 capability labels incl implementation10/debugging7/repo_exploration7; parse/empty failures0; golden and oracle-evidence source sessions used0. Main independently checked row bytes/digest/scope; old 3-session artifact remains invalid. Cross-layer checkpoint restart tests 28 OK. Full-size model observation NOT yet launched: producer is adding pre-forward all90-file original-Xiaomi digest/provenance verification and selector is validating complete HOPE/REAP grids and recomputing agreement before any selection. Next: focused source-integrity tests + final go, then all-layer pass.
